app/single_resource/views.py

```diff
@@ -45,24 +57,102 @@ def create():
 def edit(resource_id):
     """Edit a resource."""
     resource = Resource.query.get(resource_id)
-    field_names = Resource.__table__.columns.keys()
+    if resource is None:
+        abort(404)
+    resource_field_names = Resource.__table__.columns.keys()
+    descriptors = Descriptor.query.all()
+    for descriptor in descriptors:
+        if (descriptor.values):  # fields for option descriptors
+            choices = [(str(i), v) for i, v in enumerate(descriptor.values)]
+            default = None
+            option_association = OptionAssociation.query.filter_by(
+                resource_id=resource_id,
+                descriptor_id=descriptor.id
+            ).first()
+            if (option_association is not None):
+                default = option_association.option
+            setattr(SingleResourceForm,
+                    descriptor.name,
+                    SelectField(choices=choices, default=default))
+        else:  # fields for text descriptors
+            default = None
+            text_association = TextAssociation.query.filter_by(
+                resource_id=resource_id,
+                descriptor_id=descriptor.id
+            ).first()
+            if (text_association is not None):
+                default = text_association.text
+            setattr(SingleResourceForm,
+                    descriptor.name,
+                    StringField(default=default))
     form = SingleResourceForm()
     if form.validate_on_submit():
-        resource.name = form.name.data
-        resource.address = form.address.data
-        resource.latitude = form.latitude.data
-        resource.longitude = form.longitude.data
+        for field_name in resource_field_names[1:]:
+            setattr(resource, field_name, form[field_name].data)
+        add_associations(resource=resource,
+                         form=form,
+                         descriptors=descriptors,
+                         resource_existed=True)
         try:
             db.session.commit()
             flash('Resource updated', 'form-success')
-            return redirect(url_for('single_resource.edit',
-                                    resource_id=resource_id))
+            return redirect(url_for('single_resource.index'))
         except IntegrityError:
             db.session.rollback()
             flash('Error: failed to save resource. Please try again.',
                   'form-error')
-    for field_name in field_names[1:]:
+    for field_name in resource_field_names[1:]:
         form[field_name].data = resource.__dict__[field_name]
     return render_template('single_resource/edit.html',
                            form=form,
                            resource_id=resource_id)
+
+
+def add_associations(resource, form, descriptors, resource_existed=True):
```

Comment:
Could you write some documentation for this function? I'm not quite sure what it does.